rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,272 @@
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps<SampleState>({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props, we set defaults internally but if consumer wants component variants with different defaults, they can set them
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps): SampleState => {
+  const resolvedRef = useMergedRefs(ref, React.useRef());
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const state = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    state.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // enlarge the 'uber' state
+  state.someState = someState;
+  state.contextValue = contextValue;
+
+  return state;
+};
+```
+
+#### useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots.
+
+```typescript
+import { makeStyles, ax } from '@fluentui/react-make-styles';
+import { SampleState } from './Sample.types';
+
+/**
+* Styles for the root slot
+*/
+export const useRootStyles = makeStyles<SampleState>([
+  [
+    null,
+    theme => ({
+      backgroundColor: theme['select']['from']['theme']['docs'],
+      paddingRight: '12px',
+    }),
+  ],
+]);
+
+/**
+* Styles for the icon slot, uses state selectors from SampleState
+*/
+export const useIconStyles = makeStyles<SampleState>([
+  [
+    // Conditionally apply styles
+    (someState) => someState == 1 ? true : false,
+    () => ({
+      width: '20px',
+      height: '20px',
+      marginRight: '9px',
+    }),
+  ],
+]);
+
+/**
+* Applies style classnames to slots
+*/
+export const useSampleStyles = (state: SampleState) => {
+  const rootClassName = useRootStyles(state);
+  const iconClassName = useIconStyles(state);
+
+  // ax is a util that deduplicates classnames
+  state.className = ax(rootClassName, state.className);
+
+  if (state.icon) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    // currently shorthand slot typings are broken
```

Comment:
```suggestion
```

Is it still needed?

Reply:
nope, but was only true as of 5min ago 😋